accord-core/src/main/java/accord/local/Command.java

```diff
@@ -380,119 +402,155 @@ public void onChange(Command command)
+                notifyListeners();
                 break;
             case Executed:
-                writes.apply(commandStore);
-                status = Applied;
-                if (notifyListeners)
-                    listeners.forEach(this);
+                logger.trace("{}: applying", txnId());
+                if (notifyListenersOnNoop) notifyListeners();
+                return apply();
         }
+        return Writes.SUCCESS;
     }
 
     /**
      * @param dependency is either committed or invalidated
      */
-    private void updatePredecessor(Command dependency)
+    private void updatePredecessor(PartialCommand dependency)
     {
         Preconditions.checkState(dependency.hasBeen(Committed));
         if (dependency.hasBeen(Invalidated))
         {
+            logger.trace("{}: {} is invalidated. Stop listening and removing from waiting on commit set.", txnId(), dependency.txnId());
             dependency.removeListener(this);
-            if (waitingOnCommit.remove(dependency.txnId) != null && waitingOnCommit.isEmpty())
-                waitingOnCommit = null;
+            removeWaitingOnCommit(dependency);
         }
-        else if (dependency.executeAt.compareTo(executeAt) > 0)
+        else if (dependency.executeAt().compareTo(executeAt()) > 0)
         {
             // cannot be a predecessor if we execute later
+            logger.trace("{}: {} executes after us. Stop listening.", txnId(), dependency.txnId());
             dependency.removeListener(this);
         }
         else if (dependency.hasBeen(Applied))
         {
-            waitingOnApply.remove(dependency.executeAt);
+            logger.trace("{}: {} has been applied. Stop listening and removing from waiting on apply set.", txnId(), dependency.txnId());
+            removeWaitingOnApply(dependency);
             dependency.removeListener(this);
         }
         else
         {
-            waitingOnApply.putIfAbsent(dependency.executeAt, dependency);
+            logger.trace("{}: adding {} to waiting on apply set.", txnId(), dependency.txnId());
+            addWaitingOnApplyIfAbsent(dependency);
         }
     }
 
     // TEMPORARY: once we can invalidate commands that have not been witnessed on any shard, we do not need to know the home shard
     static class BlockedBy
     {
-        final TxnId txnId;
+        final PartialCommand command;
         final Keys someKeys;
 
-        BlockedBy(TxnId txnId, Keys someKeys)
+        BlockedBy(PartialCommand command, Keys someKeys)
         {
-            this.txnId = txnId;
+            this.command = command;
             this.someKeys = someKeys;
         }
     }
 
     public BlockedBy blockedBy()
     {
         Command prev = this;
-        Command cur = directlyBlockedBy();
+        PartialCommand cur = directlyBlockedBy();
         if (cur == null)
             return null;
 
-        Command next;
```

Comment:
Removing this loop breaks the progress log. I can see why it has been done, but we'll have to do a bit more work here.

Reply:
This was the change i was trying to validate with the burn test, but it was already hanging on the progress log. Sorry if I forgot to explicitly call this change out, but yeah, we can’t pointer chase in C*